### ISSUE TITLE
Octopus Tracker Energy Pricing

### ISSUE THREAD

**Issue by acos0874** [RainbowMiner/RainbowMiner#2080]
Good evening,

Since my electricity price is up and my mining revenue is down, I'm nearly at a point of breaking even, which means not mining any more. Octopus Energy in the UK offer a tariff which tracks the whole sale costs. I have been thinking about switching to it since then I could mine on some days when electricity is cheap and turn off my rigs on other days where the electricity is expensive. 

They release their pricing each day in advance so I could potentially enter the price for the following day into RBM using the scheduler, but would it be possible please to automate this somehow in RBM? I believe they have an API since they support the use of IFTTT, but it's beyond my technical knowledge.

Here are some useful sites:

https://energy.guylipman.com/sm/electracker

https://octopus.energy/tracker/

https://twitter.com/energystatsuk

**Comment by RainbowMiner** [comment]
Hi, if there is a public API, then it's pretty easy to implement. I will look for more information. If you tend to sign up with them (or already have), I suppose you directly ask them about an API, so that you can smoothly track your current rate.

**Comment by acos0874** [comment]
That would be great thanks. I'm waiting for a response from their customer
service team.

**Comment by acos0874** [comment]
Hi, I've now signed up for the tracker tariff and it's active. I haven't had much luck finding out about an API but I know some people do it. On this website my tariff appears to be region B.#

https://energy.guylipman.com/sm/electracker

Any help with this would be much appreciated.

**Comment by RainbowMiner** [comment]
Do you have any other tariff info/data/descriptor? I cannot find your tariff, at the moment. This is the data provided via the API
code                        | display_name                           
----                        | ------------                           
AFFECT-FIX-12M-22-04-21     | Affect 12M Fixed                       
AFFECT-OCC-VAR-21-10-01     | Affect Occupier Standard Tariff        
AFFECT-SEG-FIX-12M-20-11-11 | Affect Smart Export Guarantee          
AFFECT-VAR-22-04-02         | Affect Standard Tariff                 
AGILE-18-02-21              | Agile Octopus                          
AGILE-OUTGOING-19-05-13     | Agile Outgoing Octopus                 
COOP-FIX-12M-22-04-21       | Co-op 12M Fixed                        
COOP-LOYAL-FIX-12M-22-04-21 | Co-op Loyal 12M Fixed                  
COOP-OCC-VAR-21-10-01       | Co-op Occupier Flexible                
COOP-PP-VAR-20-04-01        | Co-op Key and Card                     
COOP-SEG-FIX-12M-20-11-11   | Co-op Smart Export Guarantee           
COOP-VAR-22-04-02           | Co-op Flexible                         
CP-12M-22-04-21             | Co-op Community Power                  
EB-PP-VAR-21-10-08          | Ebico Prepayment                       
ENERGY-LOCAL-LONDON         | Energy Local London                    
EP-FIX-12M-22-04-21         | Ebico Prime 12 Saver                   
ES-OCC-VAR-21-10-01         | Ebico Occupier Standard                
ES-VAR-22-04-02             | Ebico Standard                         
GO-22-03-29                 | Octopus Go                             
GO-4H-0130-22-03-29         | Octopus Go Faster                      
GO-4H-0230-22-03-29         | Octopus Go Faster                      
GO-4H-2130-22-03-29         | Octopus Go Faster                      
GO-4H-2230-22-03-29         | Octopus Go Faster                      
GO-4H-2330-22-03-29         | Octopus Go Faster                      
GO-5H-0030-22-03-29         | Octopus Go Faster                      
GO-5H-0130-22-03-29         | Octopus Go Faster                      
GO-5H-2130-22-03-29         | Octopus Go Faster                      
GO-5H-2230-22-03-29         | Octopus Go Faster                      
GO-5H-2330-22-03-29         | Octopus Go Faster                      
GO-GREEN-22-03-29           | Octopus Go Green                       
INTELLI-22-03-29            | Intelligent Octopus                    
LOYAL-FIX-12M-22-04-21      | Loyal Octopus 12M Fixed                
LP-CN-FIX-12M-22-04-21      | my london carbon neutral plan          
LP-CNT-FIX-12M-22-04-21     | my london tree plan                    
LP-HOME-21-10-01            | my london new home flex plan           
LP-SEG-FIX-12M-20-11-11     | my london smart export guarantee       
M-AND-S-12M-22-03-25        | M&S 12M Fixed                          
M-AND-S-SEG-FIX-12M-20-11-11 | M&S Smart Export Guarantee             
M-AND-S-VAR-22-04-02        | M&S Everyday Energy                    
OCC-VAR-21-10-01            | Occupier Flexible Octopus              
OE-FIX-12M-22-04-21         | Octopus 12M Fixed                      
OUTGOING-FIX-12M-19-05-13   | Outgoing Octopus 12M Fixed             
OUTGOING-SEG-FIX-12M-20-07-07 | Octopus Outgoing Smart Export Guarantee
PREPAY-VAR-18-09-21         | Octopus Key and Card                   
SUPER-GREEN-12M-22-04-21    | Super Green Octopus                    
VAR-22-04-02                | Flexible Octopus

**Comment by RainbowMiner** [comment]
Sure, here you are:

`https://api.octopus.energy/v1/products/{PRODUCT_CODE}/`
e.g.
`https://api.octopus.energy/v1/products/VAR-22-04-02/`

**Comment by RainbowMiner** [comment]
Maybe, you have some tariff code somewhere. It's build like `E-1R-VAR-22-04-02-B` - with that, one can directly call in the current rates.

**Comment by acos0874** [comment]
G-1R-SILVER-22-04-25-B

**Comment by acos0874** [comment]
E-1R-ES-VAR-22-04-02-B

**Comment by acos0874** [comment]
I can't get any combination of these to work.

**Comment by acos0874** [comment]
Gas Tracker
The Gas tracker prices are not currently available from the main Octopus API. They are available from an alternative API, but I haven't had a chance to use these - if you want to understand this you are best to ask on the forum.

Ha I just found this. So it exists but I need to find it.

**Comment by RainbowMiner** [comment]
But this tariff code `E-1R-ES-VAR-22-04-02-B` works - the product code is ES-VAR-22-04-02

**Comment by acos0874** [comment]
The key must be hidden somewhere in this page...

https://energy.guylipman.com/sm/electracker

Since the value circled exactly matches my tariff. I can't see in the source code for the page where the data is drawn in from though.

![image](https://user-images.githubusercontent.com/42582444/172670233-1deb1920-10d5-4ef8-a7e2-f7007a052607.png)

**Comment by acos0874** [comment]
Oo that worked. Also gives the data for the next day ahead...

https://api.octopus.energy/v1/products/SILVER-2017-1/electricity-tariffs/E-1R-SILVER-2017-1-B/standard-unit-rates/

**Comment by RainbowMiner** [comment]
And are you 100% sure that you have booked the correct electricity tariff? Where did you get the other tariff code E-1R-ES-VAR-22-04-02-B from?

**Comment by acos0874** [comment]
I used the guylipman website to look up my account details. The standard variable was what I was on before selecting the tracker tariff. Although my tariff code is different, I'm confident it is the same unit rate as it's the same region but with a different start date.

![image](https://user-images.githubusercontent.com/42582444/172795236-6df399ae-8156-4693-9729-5104d9156aab.png)

**Comment by acos0874** [comment]
Amazing thank you. I will send you a tip 👍

**Comment by acos0874** [comment]
Hi I have entered the tariff info..

  "OctopusTariffCode": "E-1R-SILVER-2017-1-B"

But it it hard to tell if it is working. My current unit price is 20.895 but the config is still saying this...

  "PowerPrice": "0.2402",

Is there something else I have to change? I already tried restarting rbm. I'm currently using the scheduler but I wouldn't have thought that would interfere with the main config file.
[debug_2022-06-12.zip](https://github.com/RainbowMiner/RainbowMiner/files/8885658/debug_2022-06-12.zip)

**Comment by acos0874** [comment]
Ah I figured it out. It uses the tariff prices but doesn't overwrite the powerprice value. Thanks again.

**Comment by RainbowMiner** [comment]
> Ah I figured it out. It uses the tariff prices but doesn't overwrite the powerprice value. Thanks again.

Yes, totally correct. I have replaced all references to the config's PowerPrice with a new "Get-PowerPrice" function. That function decides which value to use, of course depending on the config.

**Comment by acos0874** [comment]
Hi, sorry to reopen this. I switched to a new tariff, which is the only in the URL here.

https://octopus.energy/tracker-demo-detail/?context=1&code=E-1R-SILVER-FLEX-22-11-25-B&consumption=6103&type=electricity&inapp=true

When I pass the tariff code E-1R-SILVER-FLEX-22-11-25-B into the RainbowMiner config it won't pick up the correct price. I have looked in the API and can't see anything useful but I'm not very experienced with it. Would you mind having a look to see if it's possible for me to track this tariff code please?

**Comment by RainbowMiner** [comment]
Don't worry, I will figure this out.

**Comment by samdave123** [comment]
I've just started using this, but I'm confused whether it's working or not. The power price is still the same as before. How do I verify that RBM is actually changing the power price every 30 minutes?
